From b7cd8c618efe1daa574987b688ef3ead20348ec8 Mon Sep 17 00:00:00 2001
From: Ivan Vecera <ivecera@redhat.com>
Date: Mon, 8 Sep 2014 16:13:39 -0400
Subject: [ethernet] be2net: Fix vlans_added counter

Message-id: <1410192881-30702-17-git-send-email-ivecera@redhat.com>
Patchwork-id: 92402
O-Subject: [RHEL7.1 PATCH 16/78] be2net: Fix vlans_added counter
Bugzilla: 1088575
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>

BZ: #1088575

Upstream commit(s):
commit 48291c22b75adbbd15227070088c761c04e48a3b
Author: Vasundhara Volam <vasundhara.volam@emulex.com>
Date:   Tue Mar 11 18:53:08 2014 +0530

    be2net: Fix vlans_added counter

    When a VLAN is added by user, adapter->vlans_added is incremented.
    But if the VLAN is already programmed in HW, driver ends up
    incrementing the counter wrongly.

    Increment the counter only if VLAN is not already programmed in the HW.

    Signed-off-by: Vasundhara Volam <vasundhara.volam@emulex.com>
    Signed-off-by: Sathya Perla <sathya.perla@emulex.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Ivan Vecera <ivecera@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/emulex/benet/be_main.c b/drivers/net/ethernet/emulex/benet/be_main.c
index 1b4dee4..181ab5c 100644
--- a/drivers/net/ethernet/emulex/benet/be_main.c
+++ b/drivers/net/ethernet/emulex/benet/be_main.c
@@ -1138,7 +1138,10 @@ static int be_vlan_add_vid(struct net_device *netdev, __be16 proto, u16 vid)
 
  /* Packets with VID 0 are always received by Lancer by default */
  if (lancer_chip(adapter) && vid == 0)
-  goto ret;
+  return status;
+
+ if (adapter->vlan_tag[vid])
+  return status;
 
  adapter->vlan_tag[vid] = 1;
  adapter->vlans_added++;
@@ -1148,7 +1151,7 @@ static int be_vlan_add_vid(struct net_device *netdev, __be16 proto, u16 vid)
   adapter->vlans_added--;
   adapter->vlan_tag[vid] = 0;
  }
-ret:
+
  return status;
 }
 
-- 
1.7.1